src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala

```diff
@@ -1,46 +1,70 @@
 import org.broadinstitute.dsde.workbench.model.google.ServiceAccountSubjectId
-import org.broadinstitute.dsde.workbench.model.{WorkbenchEmail, WorkbenchUserId}
+import org.broadinstitute.dsde.workbench.model.{WorkbenchEmail, _}
 
 import scala.concurrent.{ExecutionContext, Future}
+import StandardUserInfoDirectives._
+import org.broadinstitute.dsde.workbench.sam.directory.DirectoryDAO
 
+import scala.util.Try
 
 
 trait StandardUserInfoDirectives extends UserInfoDirectives {
   implicit val executionContext: ExecutionContext
 
+  def requireUserInfo: Directive1[UserInfo] = (
+    headerValueByName(accessTokenHeader) &
+      headerValueByName(googleSubjectIdHeader) &
+      headerValueByName(expiresInHeader) &
+      headerValueByName(emailHeader)
+    ) tflatMap {
+    case (token, googleSubjectId, expiresIn, email) =>
+      onSuccess(Try(expiresIn.toLong).fold[Future[UserInfo]](t => Future.failed(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"expiresIn $expiresIn can't be converted to Long"))),
+        l => createUserInfo(OAuth2BearerToken(token), GoogleSubjectId(googleSubjectId), WorkbenchEmail(email), l, directoryDAO)))
+  }
+
+  def requireCreateUser: Directive1[CreateWorkbenchUserAPI] = (
+      headerValueByName(googleSubjectIdHeader) &
+      headerValueByName(emailHeader)
+    ) tflatMap {
+    case (googleSubjectId, email) => {
+      onSuccess(createWorkbenchUserAPI(GoogleSubjectId(googleSubjectId), WorkbenchEmail(email)))
+    }
+  }
+}
+
+object StandardUserInfoDirectives{
   val petSAdomain = "\\S+@\\S+\\.iam\\.gserviceaccount\\.com".r
 
-  private def isPetSA(email: WorkbenchEmail) = {
+  def isPetSA(email: WorkbenchEmail) = {
     petSAdomain.pattern.matcher(email.value).matches
   }
 
-  def requireUserInfo: Directive1[UserInfo] = (
-    headerValueByName("OIDC_access_token") &
-      headerValueByName("OIDC_CLAIM_user_id") &
-      headerValueByName("OIDC_CLAIM_expires_in") &
-      headerValueByName("OIDC_CLAIM_email")
-    ) tflatMap {
-    case (token, userId, expiresIn, email) => {
-      val userInfo = UserInfo(OAuth2BearerToken(token), WorkbenchUserId(userId), WorkbenchEmail(email), expiresIn.toLong)
-      onSuccess(getUserFromPetServiceAccount(userInfo).map {
-        case Some(petOwnerUser) => UserInfo(OAuth2BearerToken(token), petOwnerUser.id, petOwnerUser.email, expiresIn.toLong)
-        case None => userInfo
-      })
-    }
+  def createUserInfo(token: OAuth2BearerToken, googleSubjectId: GoogleSubjectId, email: WorkbenchEmail, expiresIn: Long, directoryDAO: DirectoryDAO)(implicit ec: ExecutionContext): Future[UserInfo] = if (isPetSA(email)) {
+    // If it's a PET account, we treat it as its owner
+    directoryDAO.getUserFromPetServiceAccount(ServiceAccountSubjectId(googleSubjectId.value)).flatMap(_.fold[Future[UserInfo]](Future.failed(new WorkbenchException("no owner found for this PET account")))(pet => Future.successful(UserInfo(token, pet.id, pet.email, expiresIn.toLong))))
```

Comment:
The idea is so that we're not throwing exceptions. For `fold`, I can change it to pattern match, but `fold` works more naturally when I'm writing out this code. For your confusion, empty case always comes first, it's the same way when you use `fold` on `Either`, `Validated` etc (failure is the first function)